warehouse/macaroons/caveats.py

```diff
@@ -24,9 +25,16 @@ class InvalidMacaroonError(Exception):
 class Caveat:
     def __init__(self, verifier):
         self.verifier = verifier
+        self.failure_reason = None
+
+    def _fail(self, reason):
+        # TODO: Surface this failure reason to the user.
+        # See: https://github.com/pypa/warehouse/issues/9018
+        self.failure_reason = reason
+        return False
 
     def verify(self, predicate):
-        raise InvalidMacaroonError
+        return self._fail("programming error")
```

Comment:
I'm not sure I love this pattern, for two reasons:

- It's not hard to imagine someone calling `self._fail(...)` (and expecting that call to change the flow) instead of `return self._fail(...)`, accidentally returning the implicit `None` instead. 
- The fact that `self._fail` essentially wraps setting state and just always returns `False` seems unnecessary.

I think I'd prefer sticking with an exception-driven control flow but I'm not sure I understand why you opted to move aware from it here.

Reply:
I'm also not in love with this pattern, but the original (exception-driven) control flow was actually masking a bug in caveat verification (one that we never experienced, since we've only ever had one caveat per macaroon before).

Using exceptions here was actually causing us to prematurely fail our caveat verification steps, since `pymacaroons` runs each verifier once per caveat body. This meant that the `V1Caveat` would raise as soon as it was passed an `ExpiryCaveat` body, failing the entire verification phase rather than returning control to the next verifier.

IMO we need _some_ kind of state handler to make the error messages returned for macaroons more useful (per #9018), although it definitely doesn't need to be this particular `_fail` thing I just made up :slightly_smiling_face:

Reply:
Got it, that makes sense.

As a compromise, can we just do something like this instead::

```suggestion
        self.failure_reason = "programming error"
        return False
```

and add some static typing to ensure that `verify` always returns a boolean?

Reply:
Yep, SGTM. I'll make these changes now.

Reply:
[e0d1709](https://github.com/pypa/warehouse/pull/11122/commits/e0d1709a1898b9c1934f7c9860d6ed2c5daa468b)